Vision Service: Optimize Docker container image

sgflt opened the pull request:
### Description

 This PR reduces size produced image and introduces lazy initialization:

- Layers are ordered from the layer that changes the least to the alyer that is changed almost in every  build.
- Dockerfile now contains volumes as suggestion where to mount folders from host, because they will be really big.
- Set gunicorn as production grade server instead of flask development server.

#### Related Issues

- #6

### Acceptance Criteria

- [x] New features or enhancements are fully implemented and do not break existing functionality, so that they can be released at any time without requiring additional work
- [x] Documentation has been / will be updated, especially as it relates to new configuration options or potentially disruptive changes

---

### Contribution Agreement

*PhotoPrism UG ("PhotoPrism", "we" or "us") hereby confirms that, to the fullest extent permitted by applicable law, this Contribution is provided "AS IS" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, ANY IMPLIED WARRANTIES OR CONDITIONS OF NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE. You have no obligation to provide support, maintenance, or other services for your Contribution.*

If you contribute code or other intellectual property, we kindly ask you to also confirm the following, so that we can safely use it in our projects without risking unexpected legal disputes:

- [x] I grant PhotoPrism a non-exclusive, perpetual, irrevocable, worldwide, fully-paid, royalty-free, and transferable right to use, copy, modify, merge, publish, distribute, sublicense and/or sell my Contributions without any restrictions. This includes a non-exclusive, perpetual, irrevocable, worldwide, fully-paid, royalty-free, and transferable patent license to use, offer for sale, sell, import, export, and otherwise transfer or make available my Contributions. PhotoPrism may, in its sole discretion, apply any license to my Contributions that it deems appropriate for the particular purpose, including other open source, copyleft, and proprietary licenses. PhotoPrism may assign this Agreement and all of its rights, obligations and licenses hereunder.
- [x] I confirm that I am the sole author of this Contribution and am legally authorized to grant the above licenses and waivers with respect to this Contribution and any other of my Contributions. If your Contributions were created in the course of your employment with your former or current employer, you represent that this employer has authorized you to make your Contributions on its behalf or has waived all rights, claims, or interests in your Contributions.

Thank you very much!

sgflt committed:
#6 Added gunicorn as production server

sgflt committed:
#6 Fixed name of hf_xet package

sgflt committed:
#6 Optimized dockerfile

- run server as non-root user
- added volume markers for large folders
- run gunicorn in production
- removed nodejs and go variables, there is no go nor nodejs
- moved application files to the last layer as they will be changed more often
- delayed venv initialization to first start of container
- made scripts non writable

sgflt committed:
#6 Fixed newlines at the end of files

lastzero commented:
@sgflt Thank you very much! I've done a (very) quick review of the changes and have the following questions:

1. From what I could see, you replaced `flask` with `gunicorn`, or is this only in production mode? I work mostly in Go, so it would be interesting to know what the advantages are (and if there are any disadvantages or things to keep in mind from an end-user/developer perspective)?
2. To run the container/service as a non-root user, it looks like you configured the username and ID with build ARGs, so once the image is pushed to Docker Hub, it can't be easily changed (at least not with environment variables)? This may not be flexible enough for some users, especially if they want to use GPU acceleration and UID 1000 doesn't have permission to use the hardware device...

In general, you could use both `ARG` and `ENV` in the `Dockerfile` to allow overriding build arguments with runtime values, as shown in this example from [vsupalov.com/docker-arg-env-variable-guide/](https://vsupalov.com/docker-arg-env-variable-guide/) (you should also find examples for this in our main repository):

```Dockerfile
# expect a build-time variable
ARG A_VARIABLE
# use the value to set the ENV var default
ENV an_env_var=$A_VARIABLE
```

To avoid confusion about environment variable names, it would be helpful if the names documented here also worked for setting e.g. the server UID and GID, at least in addition to any other variants:

- https://docs.photoprism.app/getting-started/config-options/#docker-image

In a perfect world, changing or adding the `user:` service property in your `compose.yaml` file (or specifying the corresponding `docker` command flag) should be enough to set the process UID and GID, but apparently there are NAS devices and Docker UIs that don't support/allow this. So we decided to (also) support configuration with environment variables.

lastzero commented:
Please note that I've updated the issue description to match our updated PR template. Feel free to edit the description and/or acceptance criteria as needed!

sgflt commented:
ad 1) after start there was a message ` INFO - WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.` And gunicorn is that WSGI. There is also uvicorn as ASGI (async), but for FastAPI instead of FLask.

ad 2) ok, i will remove that

sgflt committed:
#6 Added default user configuration to compose files

sgflt committed:
#6 Made uid and gid as environment variables

sgflt committed:
#6 Removed unrelated comment

sgflt committed:
#6 Use variable home destination

sgflt committed:
#6 Added camera emoji to prompt

sgflt committed:
#6 Removed default user from Dockerfile

- some old NAS container implementations may have problems with changing of default user id

sgflt committed:
#6 Use gosu for user id switching

- note that we have to pass user id to exec command too: podman exec --user 1000
- podman/docker exec without a user parameter will run as root

lastzero commented on service/scripts/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+/app/scripts/requirements.sh
```
@sgflt  As I understand it, these changes will **defer the installation of the requirements** (Python dependencies) **until the container is first started**, resulting in a **smaller image size in exchange for a longer startup time** because the packages have to be downloaded first (which is why `/app/venv` was added as a named volume in the production `compose.yaml` file)?

If so, that's certainly something we can consider... it would still be important for me to get a full understanding of the (potential) impact for us and our users:

1. What if one of the dependencies is **temporarily unavailable** (e.g. because the Python package index or Internet connection is down), or there is a newer version that hasn't been tested yet?
2. Couldn't this **break things**, or worse, **install malicious code with the packages**, e.g. if the maintainer's account has been hacked in the meantime? :unlock:

**For context, see the following AI overview from Google:**

> Malicious Python packages can be uploaded to the Python Package Index (PyPI) and contain code that steals data, installs remote access tools (RATs), or performs other harmful actions. These packages can be disguised as legitimate updates or fixes, making them difficult to detect.
>
> **Be cautious about updating packages:** Verify the package name and author before updating. Consider using virtual environments to isolate package installations. 

![python](https://github.com/user-attachments/assets/9b25e7e3-f04c-4d8c-9a17-27d36527e453)

sgflt replied:
Yes, I was thinking about locking version in requirements.txt. But it cannot make anything with supply chain attacs. 

Does github support some type of scanning like snyk for python? It is probably the only thing that is doable.

sgflt replied:
Pushed current versions.

lastzero commented on compose.yaml:
```diff
@@ -41,6 +42,10 @@ services:
               count: "all"
+    volumes:
+      - models:/app/models
+      - venv:/app/venv
```
It could make sense to specify the `/app/venv` and `/app/models` directories as `VOLUME` in the `Dockerfile`:

- https://docs.docker.com/reference/dockerfile/#volume

This way, **(a)** they are initialized with the data already present in the base image, and **(b)** users (presumably) don't need to explicitly configure them in their `compose.yaml` file, which can simplify things and **(c)** avoids reinstalling the dependencies after every configuration change or update if the volume was not specified in the `compose.yaml` file for some reason.

sgflt replied:
It already is. Isn't it?

lastzero replied:
It seems you're right... I probably didn't notice it because the `VOLUME` instructions are combined with `--chown=` - is that really possible? :thinking: 

I know you can use it with `ADD`, but I haven't seen it documented/used with `VOLUME` yet.

sgflt replied:
Interesting, you are right. It did nothing, so I've removed the chown.

sgflt committed:
#6 Locked versions of python dependencies to ensure a repeatable build and environment

sgflt committed:
#6 Removed NOOP chown from VOLUME

- it does nothing

lastzero commented:
Once you've checked the boxes in the PR description above, I'll do a final review and test so we can merge this. You can also edit the acceptance criteria as needed.

lastzero commented:
Sorry for not checking back sooner! I'll now merge your changes for final testing :+1:
